### PR TITLE
Fix functions equality

### PR DESCRIPTION
**Pull request by openorclose**
#88

### CODE REVIEW

**Comment by thomastanck** [comment]
Might want to update this PR by resolving the merge conflicts.

File: src/types.ts [hunk 1]
```diff
@@ -153,50 +153,65 @@ export interface Frame {
...
   /** Keep track how many lambdas are created */
   private static lambdaCtr = 0
 
   /** Unique ID defined for anonymous closure */
-  public name: string
+  public functionName: string
 
   /** Fake closure function */
-  // tslint:disable-next-line:ban-types
+    // tslint:disable-next-line:ban-types
```
Comment (thomastanck) [inline src/types.ts]: Unnecessary spaces?

File: src/__tests__/index.ts [hunk 2]
```diff
@@ -189,6 +203,21 @@ test(
   30000
 )
 
+test("Functions passed into non-source functions remain equal", () => {
+  const code = `
+  function t(x, y, z) {
+    return x + y + z;
+  }
+  head(pair(t, t)) === t && t(1, 2, 3) === 6;
```
Comment (thomastanck) [inline src/__tests__/index.ts]: I think to be more robust against future library changes (what if head/pair are later implemented in pure Source), you could use `defineSymbol` from `createContext.ts` to create your own built-in that's guaranteed to be non-source.

File: src/types.ts [hunk 3]
```diff
@@ -153,50 +153,65 @@ export interface Frame {
...
+    const funJS = closureToJS(this, context, this.functionName)
+    this.fun = funJS
   }
 }
 
 /**
  * Modified from class Closure, for construction of arrow functions.
  */
-export class ArrowClosure {
+export class ArrowClosure extends Callable{
```
Comment (thomastanck) [inline src/types.ts]: Missing space.

Comment (openorclose) [reply src/types.ts]: Done

File: src/__tests__/index.ts (hunk 2 above)
Comment (openorclose) [reply src/__tests__/index.ts]: Yes, it was probably wrong to rely on external "builtins".

**Pull request merged**